### PR TITLE
Apache Ignite File System

### PR DESCRIPTION
**Pull request by dmitrievanthony**
This is a proposal to support [Apache Ignite File System](https://ignite.apache.org/features/igfs.html) in TensorFlow.

[Apache Ignite](https://ignite.apache.org/) is a memory-centric distributed database, caching, and processing platform for transactional, analytical, and streaming workloads, delivering in-memory speeds at petabyte scale. This proposal is a part of a more global initiative to implement so called "TensorFlow on Apache Ignite" ([IGNITE-8335](https://issues.apache.org/jira/browse/IGNITE-8335), [Design Document](https://docs.google.com/document/d/1jROIahK1rc7bSgOvhJhfpMqIGvht_IE8zn5NAt6x8ks)).

The integration is based on [custom filesystem plugin](https://www.tensorflow.org/extend/add_filesys) from TensorFlow side and IGFS Native API from Apache Ignite side. More information about this module you can find in [README.md](https://github.com/dmitrievanthony/tensorflow/blob/apache-ignite-igfs/tensorflow/contrib/igfs/README.md).

Tests have also been added. They use docker to hide configuration complexity, so that the implemented functionality can be tested quite simply by manual run.

**Be aware that this pull request is based on previous #22210 that is not merged yet (significant part of the code you see here will be merged in #22210).**

### CODE REVIEW

**Comment by yifeif** [comment]
Thanks @dmitrievanthony. Could you check the cla for all contributors who have commits in this PR?

**Comment by dmitrievanthony** [comment]
Hi, @yifeif. Thank you for attention. Did I understand correctly from the message above says that all authors have signed CLA and the problem is only that the committer and the author of one commit are not the same? If so, I asked @artemmalykh to write here that he's ok with this PR.

**Comment by artemmalykh** [comment]
I'm totally OK with PR!

**Comment by yifeif** [comment]
Thanks @dmitrievanthony and @artemmalykh. Looks good now.

**Comment by dmitrievanthony** [comment]
@yifeif, I just rebased this branch, so commits haven't been changed. I think CLA check should be fine. I still waiting #22210 to be merged so I'll make one more rebase when it will be merged and ping you again.

**Comment by dmitrievanthony** [comment]
Hi, @mrry, @martinwicke, @yifeif.

After merge of #22210 I rebased and updated this pull request, so now it's ready for review. It's quite big, but less that Ignite Dataset. 

Derek, I'll be grateful if you have a look and start reviewing. Please let me know if changes required.

Would be great if someone starts tests and clicks `cla:yes` (cla checker is too suspicious in case of rebases).

**Comment by dmitrievanthony** [comment]
Hi, @mrry. Any updates?

File: tensorflow/contrib/igfs/kernels/igfs_client.h
Comment (mrry) [inline tensorflow/contrib/igfs/kernels/igfs_client.h]: `inline` is implied for member functions defined in the class definition.

File: tensorflow/contrib/BUILD
Comment (mrry) [inline tensorflow/contrib/BUILD]: Can we combine these two modules into one? IGFS seems like it would fit under `"//tensorflow/contrib/ignite"`.

File: tensorflow/contrib/igfs/kernels/igfs.cc
Comment (mrry) [inline tensorflow/contrib/igfs/kernels/igfs.cc]: Is the "shared"-ness ever used here? It looks like this could return a `unique_ptr`.

File: tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc
Comment (mrry) [inline tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc]: You could `std::move(key), std::move(val)` here to avoid copying the strings again.

File: tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc
Comment (mrry) [inline tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc]: This is an unsafe cast and should raise a compiler warning: `str.length()` returns a `size_t`. You need to check that it fits in the range of an `unsigned short` before casting.

File: tensorflow/contrib/igfs/kernels/igfs_messages.cc
Comment (mrry) [inline tensorflow/contrib/igfs/kernels/igfs_messages.cc]: `path` is a `const string&`, so moving it has no effect. If you want to avoid a copy, you need to take `path` as a (by-value) `string` or a `string&&`.

File: tensorflow/contrib/igfs/kernels/igfs_messages.h
Comment (mrry) [inline tensorflow/contrib/igfs/kernels/igfs_messages.h]: Suggestion: `entries.clear()` is slightly more readable, since the object must already have been constructed.

**Review by mrry — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/igfs/__init__.py
Comment (dmitrievanthony) [reply tensorflow/contrib/igfs/__init__.py]: Removed.

File: tensorflow/contrib/igfs/kernels/igfs.cc
Comment (dmitrievanthony) [reply tensorflow/contrib/igfs/kernels/igfs.cc]: Done, marked as `static`.

File: tensorflow/contrib/igfs/kernels/igfs_client.h
Comment (dmitrievanthony) [reply tensorflow/contrib/igfs/kernels/igfs_client.h]: Yes, agree. It's exactly member function defined in the class definition.

File: tensorflow/contrib/BUILD
Comment (dmitrievanthony) [reply tensorflow/contrib/BUILD]: I would prefer to have them in separate modules by several reasons. First of all, they have very different functionality (`"//tensorflow/contrib/ignite"` provides an ability to read data from Apache Ignite Cache while `"//tensorflow/contrib/igfs"` provides allows to use file system, based on Apache Ignite). User may want to use only one feature, but not other. It looks like very reasonable in some cases. I think it's a case when "write programs that do one thing and do it well" is good, isn't it? And by the way, after `contrib` sunsetting it makes even more sense to have them in separate packages.

Are you agree?

File: tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc
Comment (dmitrievanthony) [reply tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc]: Done.

File: tensorflow/contrib/igfs/kernels/igfs.cc
Comment (dmitrievanthony) [reply tensorflow/contrib/igfs/kernels/igfs.cc]: Replaced by `unique_ptr`, done.

File: tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc
Comment (dmitrievanthony) [reply tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc]: Done.

File: tensorflow/contrib/igfs/kernels/igfs_messages.cc
Comment (dmitrievanthony) [reply tensorflow/contrib/igfs/kernels/igfs_messages.cc]: Okay, just removed `copy()`.

File: tensorflow/contrib/igfs/kernels/igfs_messages.h
Comment (dmitrievanthony) [reply tensorflow/contrib/igfs/kernels/igfs_messages.h]: Done.

**Comment by dmitrievanthony** [comment]
Hi, @mrry. Thanks for review. I updated code in accordance with all your comments, please have a look.

**Comment by dmitrievanthony** [comment]
Hi @mrry. Any updates?

**Comment by dmitrievanthony** [comment]
Hi, @mrry, @yifeif. I don't clearly understand current state. Do you have some concerns about this PR or it's ready to be merged?

**Review by mrry — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc
Comment (dmitrievanthony) [reply tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc]: Good point, done.

File: tensorflow/contrib/igfs/kernels/igfs_client.h
Comment (dmitrievanthony) [reply tensorflow/contrib/igfs/kernels/igfs_client.h]: Okay, I got it. I've removed `inline` in headers here and in `ignite_client.h` and `ignite_byte_swapper.h` as well.

File: tensorflow/contrib/BUILD
Comment (dmitrievanthony) [reply tensorflow/contrib/BUILD]: Sounds reasonable. I merged two modules together.

**Comment by dmitrievanthony** [comment]
Hi @mrry. I merged `ignite` and `igfs` modules together and fixed other remarks. Please have a look.

File: tensorflow/contrib/ignite/BUILD
Comment (mrry) [inline tensorflow/contrib/ignite/BUILD]: Nit: This should be `":ignite_client"` now.

File: tensorflow/contrib/ignite/BUILD
Comment (mrry) [inline tensorflow/contrib/ignite/BUILD]: Since `":igfs_kernels"` and `":dataset_kernels"` share some symbols (in `":ignite_client"`), to avoid the ODR issues, it'll be easiest to merge "_dataset_ops.so" and "_igfs_ops.so" into a single shared object (and hence `tf_custom_op_library()` rule).

**Review by mrry — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/ignite/BUILD
Comment (dmitrievanthony) [reply tensorflow/contrib/ignite/BUILD]: Done.

File: tensorflow/contrib/ignite/BUILD
Comment (dmitrievanthony) [reply tensorflow/contrib/ignite/BUILD]: Okay, good point. I've merged them into a single shared library `_ignite_ops.so`.

**Comment by dmitrievanthony** [comment]
Hi @mrry. I've updated the code, please have a look.

**Review by mrry — DISMISSED** [review]

**Comment by dmitrievanthony** [comment]
It looks like Sanity failed because of `//tensorflow/core:lib` dependency in `igfs_kernels`. I've removed it, so please rerun tests.

**Comment by dmitrievanthony** [comment]
Thanks @mrry, looks like everything is fixed. Currently only one test fails and it fails in non related to my code part.

Hope the PR will be merged soon :)

**Pull request merged**
